### PR TITLE
Fetching brs during approval

### PR DESCRIPTION
**Pull request by aaronk**
Improved fetching of BRs when approving.  Does a batch of promises at a time.

### CODE REVIEW

File: src/components/billingRecord/IFXBillingRecordList.vue [hunk 1]
```diff
@@ -181,16 +186,14 @@ export default {
...
         return br
       }
       if (br.id) {
         // Go get it
         br = await this.apiRef.getByID(this.facility.invoicePrefix, br.id)
         this.$set(this.items, index, br)
-        console.log('the items now ', this.items)
         return br
       }
       console.log(`Billing record with id not found at item index ${index}`)
```
Comment (luthian) [inline src/components/billingRecord/IFXBillingRecordList.vue:199]: I assume this shouldn't happen but you want to log them somewhere?

File: src/components/billingRecord/IFXBillingRecordList.vue [hunk 2]
```diff
@@ -292,20 +295,24 @@ export default {
...
-          me.items.forEach((it1) => {
-            items.forEach((it2) => {
-              if (it1.id === it2.id) {
-                it1.billingRecordStates.push({ name: state, user: '', approvers: [], comment: '' })
-                toBeUpdated.push(it1)
-              }
-            })
+      const promises = missing.map((i) => me.getFullBillingRecordByItemIndex(i))
+      while (promises.length) {
+        await new Promise(r => setTimeout(r, 500));
```
Comment (luthian) [inline src/components/billingRecord/IFXBillingRecordList.vue]: Why are you batching these promises up and waiting between each? Does the backend get overwhelmed or something? Or is there supposed to be a progress bar of some kind?

File: src/components/billingRecord/IFXBillingRecordList.vue [hunk 3]
```diff
@@ -292,20 +295,24 @@ export default {
...
+        await Promise.all(promises.splice(0, this.promiseBatchSize))
+          .catch((err) => {
+            console.log('error from one of the promises ', err)
           })
-          console.log(toBeUpdated)
-          return this.$api.billingRecord.bulkUpdate(toBeUpdated, this.facility.applicationUsername)
+      }
+      const toBeUpdated = []
+      this.items.forEach((it1) => {
+        items.forEach((it2) => {
```
Comment (luthian) [inline src/components/billingRecord/IFXBillingRecordList.vue]: ~~Should this be `this.items` as above? Or is `items` defined somewhere?~~ Never mind.

File: src/components/billingRecord/IFXBillingRecordList.vue [hunk 4]
```diff
@@ -292,20 +295,24 @@ export default {
...
+        await new Promise(r => setTimeout(r, 500));
+        await Promise.all(promises.splice(0, this.promiseBatchSize))
+          .catch((err) => {
+            console.log('error from one of the promises ', err)
           })
-          console.log(toBeUpdated)
-          return this.$api.billingRecord.bulkUpdate(toBeUpdated, this.facility.applicationUsername)
+      }
+      const toBeUpdated = []
+      this.items.forEach((it1) => {
```
Comment (luthian) [inline src/components/billingRecord/IFXBillingRecordList.vue]: You might want to structure this differently to allow the loop to be broken out of once a match is found.
```suggestion
      items.forEach((it2) => {
        for (i = 0; i < this.items.length; i++) {
          if (this.items[i].id === it2.id) {
              const it1 = this.items[i]
              it1.billingRecordStates.push({ name: state, user: '', approvers: [], comment: '' })
             toBeUpdated.push(it1)
             break;
           }
         }
      }
```
Or something like that. The point is that `forEach` doesn't allow one to use `break`.

File: src/components/billingRecord/IFXBillingRecordList.vue (hunk 1 above)
Comment (aaronk) [reply src/components/billingRecord/IFXBillingRecordList.vue:199]: I think it's a developer only thing.

File: src/components/billingRecord/IFXBillingRecordList.vue (hunk 2 above)
Comment (aaronk) [reply src/components/billingRecord/IFXBillingRecordList.vue]: Mostly there would be 20-30.  A big batch of RC could be many hundreds.  Could be a bit much.

Comment (luthian) [reply src/components/billingRecord/IFXBillingRecordList.vue]: You're firing them all off at once but then waiting in batches; that's the part I don't really understand. Do you want to show a progress indicator or something between waits? Or is the `sleep` there to allow user interaction?

File: src/components/billingRecord/IFXBillingRecordList.vue [hunk 5]
```diff
@@ -292,20 +295,33 @@ export default {
...
+        for (i = 0; i < this.items.length; i++) {
+          if (this.items[i].id === it2.id) {
+              const it1 = this.items[i]
+              it1.billingRecordStates.push({ name: state, user: '', approvers: [], comment: '' })
+             toBeUpdated.push(it1)
+             break;
+           }
+         }
+      }
+        items.forEach((it2) => {
```
Comment (luthian) [inline src/components/billingRecord/IFXBillingRecordList.vue]: I can't figure out how to do a multiline comment but you should remove the duplicated lines.

**Pull request merged**